### PR TITLE
Improve preciseness of some PHPdoc annotations

### PR DESCRIPTION
**Pull request by mstilkerich**
Another few improvements to the type annotations to improve static analysis results.

### CODE REVIEW

File: program/lib/Roundcube/rcube_addressbook.php [hunk 1]
```diff
@@ -121,12 +121,12 @@ function refresh_search()
...
      *
-     * @param array  $fields   List of fields to search in
-     * @param string $value    Search value
-     * @param int    $mode     Search mode. Sum of self::SEARCH_*.
-     * @param bool   $select   True if results are requested, False if count only
-     * @param bool   $nocount  True to skip the count query (select only)
-     * @param array  $required List of fields that cannot be empty
+     * @param string|string[] $fields   List of fields to search in
+     * @param string|string[] $value    Search value
+     * @param int    $mode    Search mode. Sum of self::SEARCH_*.
```
Comment (alecpl) [inline program/lib/Roundcube/rcube_addressbook.php:126]: Please, params descriptions should be aligned with each other.

File: program/lib/Roundcube/rcube_addressbook.php [hunk 2]
```diff
@@ -121,12 +121,12 @@ function refresh_search()
...
+     * @param bool   $select  True if results are requested, False if count only
+     * @param bool   $nocount True to skip the count query (select only)
+     * @param string|string[] $required List of fields that cannot be empty
```
Comment (alecpl) [inline program/lib/Roundcube/rcube_addressbook.php:129]: The code indeed accepts a string, but I would not advertise this possibility here.

File: program/lib/Roundcube/rcube_addressbook.php [hunk 3]
```diff
@@ -121,12 +121,12 @@ function refresh_search()
     /**
      * Search records
      *
-     * @param array  $fields   List of fields to search in
-     * @param string $value    Search value
-     * @param int    $mode     Search mode. Sum of self::SEARCH_*.
-     * @param bool   $select   True if results are requested, False if count only
-     * @param bool   $nocount  True to skip the count query (select only)
-     * @param array  $required List of fields that cannot be empty
+     * @param string|string[] $fields   List of fields to search in
```
Comment (alecpl) [inline program/lib/Roundcube/rcube_addressbook.php:124]: Please, take a look at rcube_contacts param descriptions. We should use similar here. I mean the description should say something about the type difference too.

**Review by alecpl — CHANGES_REQUESTED** [review]

File: program/lib/Roundcube/rcube_addressbook.php (hunk 1 above)
Comment (mstilkerich) [reply program/lib/Roundcube/rcube_addressbook.php]: Fixed

File: program/lib/Roundcube/rcube_addressbook.php (hunk 2 above)
Comment (mstilkerich) [reply program/lib/Roundcube/rcube_addressbook.php:151]: From my viewpoint as developer of an implementation of this interface, I'm looking at this not as advertisement of possibilities but as required expected behavior that I have to implement. So unless roundcube guarantees not to use this interface with string parameters, it must be listed here.
Looking at the roundcube code, it seems to _exclusively_ use this parameter in its string form, I did not find a single use of the array variant:

```
./program/actions/mail/search_contacts.php:            $result = $source->search($afields, $search, $search_mode, true, true, 'email');
./program/actions/mail/autocomplete.php:                if ($result = $abook->search($fields, $search, $mode, true, true, 'email')) {
./program/actions/contacts/mailto.php:                $sources[] = $contacts->search($contacts->primary_key, $cid, 0, true, true, 'email');
./program/actions/contacts/photo.php:                    $result = $abook->search(['email'], $email, 1, true, true, 'photo');
```
Therefore, I would suggest to keep the interface definition as proposed.

File: program/lib/Roundcube/rcube_addressbook.php (hunk 3 above)
Comment (mstilkerich) [reply program/lib/Roundcube/rcube_addressbook.php]: I adapted the texts from rcube_contacts. As the function is not very comprehensively specified, I also added more extensive description of my understanding that I gathered from studying the roundcube code. I you are open to such documentation, I would also contribute more in this direction for other functions of this interface, which would also help me to get a common understanding / agreement on the interface definition for addressbook implementations.

**Comment by mstilkerich** [comment]
Thanks for your fast feedback, Alec. I really appreciate your responsiveness to pull requests. Please find my comments as replies to yours.

File: program/lib/Roundcube/rcube_result_set.php [hunk 4]
```diff
@@ -21,16 +21,35 @@
...
- * Implenets Iterator and thus be used in foreach() loops.
+ * Implements Iterator and can thus be used in foreach() loops.
  *
  * @package    Framework
  * @subpackage Addressbook
  */
 class rcube_result_set implements Iterator, ArrayAccess
 {
+    /**
+     * @var int $count  The number of total records. Note that when only a subset of records of records is requested,
```
Comment (alecpl) [inline program/lib/Roundcube/rcube_result_set.php:32]: Variable name here is optional, so I propose to not use it. Also, the extra spaces between type and description are not needed.

File: program/lib/Roundcube/rcube_result_set.php [hunk 5]
```diff
@@ -21,16 +21,35 @@
...
     public $first      = 0;
+
+    /**
+     * @var bool $searchonly True if the results are from an addressbook that does not support listing all records but
+     *                       requires the search function to be used.
+     */
     public $searchonly = false;
+
+    /**
+     * @var array $records The data records of he result set. May be a subset of the total records, e.g. for one page.
```
Comment (alecpl) [inline program/lib/Roundcube/rcube_result_set.php:51]: Typo: he -> the.

File: program/lib/Roundcube/rcube_result_set.php [hunk 6]
```diff
@@ -21,16 +21,35 @@
...
  * @package    Framework
  * @subpackage Addressbook
  */
 class rcube_result_set implements Iterator, ArrayAccess
 {
+    /**
+     * @var int $count  The number of total records. Note that when only a subset of records of records is requested,
+     *                  this number may be higher than the number of data records in this result set.
+     */
     public $count      = 0;
```
Comment (alecpl) [inline program/lib/Roundcube/rcube_result_set.php:35]: After adding the comment extra spaces here are not needed and should be removed.

File: program/lib/Roundcube/rcube_addressbook.php [hunk 7]
```diff
@@ -121,14 +162,36 @@ function refresh_search()
...
      *
-     * @param array  $fields   List of fields to search in
-     * @param string $value    Search value
-     * @param int    $mode     Search mode. Sum of self::SEARCH_*.
-     * @param bool   $select   True if results are requested, False if count only
-     * @param bool   $nocount  True to skip the count query (select only)
-     * @param array  $required List of fields that cannot be empty
+     * Depending on the given parameters the search() function operates in different modes (in the order listed):
+     *
+     * Mode "Direct ID search" - when $fields is either 'ID' or $this->primary_key
```
Comment (alecpl) [inline program/lib/Roundcube/rcube_addressbook.php:167]: Nicely written, but the use of word "mode" might be confused with the $mode param below.

**Review by alecpl — CHANGES_REQUESTED** [review]

File: program/lib/Roundcube/rcube_result_set.php (hunk 4 above)
Comment (mstilkerich) [reply program/lib/Roundcube/rcube_result_set.php:32]: Fixed

File: program/lib/Roundcube/rcube_result_set.php (hunk 5 above)
Comment (mstilkerich) [reply program/lib/Roundcube/rcube_result_set.php:51]: Fixed

File: program/lib/Roundcube/rcube_result_set.php (hunk 6 above)
Comment (mstilkerich) [reply program/lib/Roundcube/rcube_result_set.php:35]: I hope I understood correctly that this is about the spaces preceeding the assignment - I removed these at all places where I added comments.

File: program/lib/Roundcube/rcube_addressbook.php (hunk 7 above)
Comment (mstilkerich) [reply program/lib/Roundcube/rcube_addressbook.php:167]: Good point. I couldn't think of a suitable alternative term for mode, so I removed terming the different modes of operation (ID search etc.) as mode.

File: program/lib/Roundcube/rcube_result_set.php [hunk 8]
```diff
@@ -21,17 +21,36 @@
...
  *
  * @package    Framework
  * @subpackage Addressbook
  */
 class rcube_result_set implements Iterator, ArrayAccess
 {
-    public $count      = 0;
-    public $first      = 0;
+    /**
+     * @var int The number of total records. Note that when only a subset of records of records is requested,
```
Comment (alecpl) [inline program/lib/Roundcube/rcube_result_set.php:32]: Last thing. Duplicated "of records".

**Review by alecpl — CHANGES_REQUESTED** [review]

File: program/lib/Roundcube/rcube_result_set.php (hunk 8 above)
Comment (mstilkerich) [reply program/lib/Roundcube/rcube_result_set.php]: Thanks, fixed.

**Pull request merged**